libs/language-server/src/lib/lsp/jayvee-scope-provider.ts

```diff
@@ -68,114 +55,103 @@ export class JayveeScopeProvider extends DefaultScopeProvider {
       return EMPTY_SCOPE;
     }
 
-    const importedUris = new Set<string>();
-    this.gatherImports(jayveeModel, importedUris);
-    this.gatherBuiltins(importedUris);
+    const importedElements: AstNodeDescription[] = [];
+    importedElements.push(...this.getBuiltinElements());
+    importedElements.push(...this.getExplicitlyImportedElements(jayveeModel));
 
-    const importedDocuments = [...importedUris].map((importedUri) =>
-      this.langiumDocuments.getDocument(URI.parse(importedUri)),
-    );
+    return new MapScope(importedElements);
+  }
 
+  protected getExplicitlyImportedElements(
+    model: JayveeModel,
+  ): AstNodeDescription[] {
     const importedElements: AstNodeDescription[] = [];
-    for (const importedDocument of importedDocuments) {
+    for (const importDefinition of model.imports) {
+      const importedDocument = this.getImportedDocument(importDefinition);
       if (importedDocument === undefined) {
         continue;
       }
 
-      const publishedElements = this.availableElementsPerDocumentCache.get(
-        importedDocument.uri,
-        'exports', // we only need one key here as it is on document basis
-        () => this.getExportedElements(importedDocument),
-      );
       importedElements.push(
-        ...publishedElements.map((e) =>
-          this.descriptions.createDescription(e.element, e.alias),
+        ...this.getImportedElementsFromDocument(
+          importDefinition,
+          importedDocument,
         ),
       );
     }
-
-    return new MapScope(importedElements);
+    return importedElements;
   }
 
-  /**
-   * Gets all exported elements from a document.
-   * This logic cannot reside in a {@link ScopeComputationProvider} but should be handled here:
-   * https://github.com/eclipse-langium/langium/discussions/1508#discussioncomment-9524544
-   */
-  protected getExportedElements(document: LangiumDocument): ExportDetails[] {
-    const model = document.parseResult.value as JayveeModel;
-    const exportedElements: ExportDetails[] = [];
-
-    for (const node of AstUtils.streamAllContents(model)) {
-      if (isExportableElementDefinition(node) && node.isPublished) {
-        assert(
-          isExportableElement(node),
-          'Exported node is not an ExportableElement',
-        );
-        exportedElements.push({
-          element: node,
-          alias: node.name,
-        });
-      }
+  protected getImportedElementsFromDocument(
+    importDefinition: ImportDefinition,
+    importedDocument: LangiumDocument,
+  ): AstNodeDescription[] {
+    const publishedElement =
```

Comment:
Nitpicky but the variable sounds like one element, the function like a list of elements.